### ISSUE TITLE
Question: KFAC/EKFAC with SGLD optimizer

### ISSUE THREAD

**Issue by har07** [Thrandis/EKFAC-pytorch#7]
Thanks for publishing the source code for your KFAC/EKFAC preconditioner implementation!

I'm sorry if you think this isn't the right place to ask/discuss this. But I tried using both KFAC and EKFAC preconditioner in combination with [SGLD](https://dl.acm.org/doi/10.5555/3104482.3104568) (basically SGD + Gaussian noise) optimizer to perform [K-FAC SGLD](https://openreview.net/forum?id=ry-Se9kvG). But the result is not as expected: training loss was increasing consistently over time. I kinda stuck because this observation didn't change despite I tried to randomly increase/decrease the value or flip the boolean value of the various preconditioner's hyperparameters.

The model that I used contains [2 `Conv2d` layers and 2 `Linear` layers](https://github.com/henripal/sgld/blob/master/sgld/sgld/model.py). Below is the loss and accuracy plots with hyperparameters as follows:

- learning rate: 0.0005
- eps: 0.5
- sua: True
- pi: True
- update_freq: 5
- alpha: 1.
- constraint_norm: True

![ksgld_loss](https://user-images.githubusercontent.com/11190794/104117653-8befb480-5355-11eb-8438-912938b152bb.png)
![ksgld_accuracy](https://user-images.githubusercontent.com/11190794/104117658-90b46880-5355-11eb-960b-3ebf97ec936e.png)

Changing the optimizer to classic SGD without changing the preconditioner's hyperparameters worked incredibly well though. It achieved 99.99% accuracy on MNIST after just 3 epochs.

Any suggestion on how to tweak the hyperparameters or to which direction? Or do you think this way of preconditioning SGLD using KFAC/EKFAC won't work for some reason?

**Comment by Thrandis** [comment]
Hey,
So I don't know anything about SLGD so it is hard to help you.

Concerning the tuning of the hyper-parameters, one thing that might help you is that when you can see `eps` as a trade-off between SGD updates and Natural Gradient updates. So the higher the `eps` the closer to SGD the updates will be. So you might want to use KFAC + SGLD with a very high `eps` to have similar steps to `SGLD` and then explore the effect of decreasing it.

**Comment by har07** [comment]
Thanks a lot for replying!

the value of `eps` should be in range between [0,1] right?

I think I have read your comment on the intuition of `eps` from the other github issue, and probably have tried `eps=1` as well, but the result wasn't any better. But maybe I just messed some other parameters while trying that. So, I'll try again to make sure, maybe with `eps=0.99`

**Comment by Thrandis** [comment]
It can be anything as long as it is positive, so it is not limited in the
range 0-1

**Comment by har07** [comment]
I have tried a few large `eps` i.e 100, 1000, and 10000 for 10 epoch. The accuracy and loss didn't seem to change even compared to `eps=0.5` that I initially used. Anyways, thanks for the information about `eps` value range!

![image](https://user-images.githubusercontent.com/11190794/104146826-a5990680-53fe-11eb-9a32-334f65aae313.png)
![image](https://user-images.githubusercontent.com/11190794/104146830-ab8ee780-53fe-11eb-984c-7026cf49cd98.png)
![image](https://user-images.githubusercontent.com/11190794/104147084-89499980-53ff-11eb-8cd6-29a983c1546d.png)

**Comment by har07** [comment]
Though the effect of changing `eps` seem to be consistent with the intuition as you explained when using KFAC with SGD optimizer

![image](https://user-images.githubusercontent.com/11190794/104178572-2a0a7a00-543d-11eb-86a1-050512bdcf0a.png)
![image](https://user-images.githubusercontent.com/11190794/104178582-2d9e0100-543d-11eb-9742-e02c69b93a22.png)

**Comment by har07** [comment]
I got an idea! I think the problem is because preconditioning applied only to the gradient while it should've been applied to both gradient and noise in case of SGLD. From what I understand, I need to multiply the approximate inverse of Fisher <img src="https://render.githubusercontent.com/render/math?math=\widetilde{F}^{-1}"> with the noise <img src="https://render.githubusercontent.com/render/math?math=\epsilon">:

![image](https://user-images.githubusercontent.com/11190794/105792552-3b659180-5fba-11eb-883c-a08b4f1ebe4c.png)

*) this version of parameter update formula in KSGLD is taken from https://arxiv.org/abs/1806.02855

From what I understand, KFAC/EKFAC calculate some approximation (block-diagonal approximation) of the inverse Fisher, then use it to precondition the gradient, right?
So do you have suggestion on where in KFAC preconditioner code that the approximate inverse Fisher is ready for use? 
I tried to look at the code but couldn't firmly point on where calculation of approximate inverse Fisher is done exactly

Sorry for bothering you again. I've been grateful for your help so far!

**Comment by tfjgeorge** [comment]
Our preconditioner works by :

1. taking the gradient in the parameter gradients (see
https://github.com/Thrandis/EKFAC-pytorch/blob/master/ekfac.py#L94 )
2. multiplying with the inverse Fisher
3. putting back the natural gradient in the parameter gradients (see
https://github.com/Thrandis/EKFAC-pytorch/blob/master/ekfac.py#L111)

In order to have the preconditioner also applied to your gaussian noise, I
recommend you add it to the parameter gradient before calling the
preconditioner.

This means:
1. Calling loss.backward() (this populates all param.grad gradients)
2. Looping through all parameters and adding gaussian noise with
param.grad.data.add_(epsilon)
3. Calling precond.step()
4. Calling optimizer.step()

Thomas
